### ISSUE TITLE
Failed to import onnx model created with keras (resnet18)

### ISSUE THREAD

**Issue by axeldavy** [onnx/onnx-tensorrt#363]
Importing a keras model exported with keras-onnx fails.

I have the following error:
While parsing node number 0 [Transpose -> "TFNodes/model_3/conv1/convolution__264:0"]:
ERROR: /home/axel/tools/TensorRT/parsers/onnx/builtin_op_importers.cpp:3205 In function importTranspose:
[8] Assertion failed: transposeWeights(weights, perm, &new_weights)

The first node of the graph is:
  node {
    input: "TFNodes105/conv1/kernel:0"
    output: "TFNodes/model_3/conv1/convolution__264:0"
    name: "TFNodes/model_3/conv1/convolution__264"
    op_type: "Transpose"
    attribute {
      name: "perm"
      ints: 3
      ints: 2
      ints: 0
      ints: 1
      type: INTS
    }
    domain: ""
  }

In other words, keras asks some transposition because it uses hwc order instead of chw order, and onnx-tensorrt fails to support it.

**Comment by axeldavy** [comment]
For anyone having a similar issue, the following workaroung on the keras side works:

Replace:
`onnx_model = keras2onnx.convert_keras(model, model.name)`

by
```
graph_def = keras2onnx.export_tf_frozen_graph(model)
onnx_model = keras2onnx.convert_tensorflow(graph_def, **keras2onnx.build_io_names_tf2onnx(model))
```

It reverts to using tf2onnx completly for the Keras model, which uses a different codepath and different optimizations which seem to make tensorrt happier.

**Comment by Toku11** [comment]
axel

> For anyone having a similar issue, the following workaroung on the keras side works:
> 
> Replace:
> `onnx_model = keras2onnx.convert_keras(model, model.name)`
> 
> by
> 
> ```
> graph_def = keras2onnx.export_tf_frozen_graph(model)
> onnx_model = keras2onnx.convert_tensorflow(graph_def, **keras2onnx.build_io_names_tf2onnx(model))
> ```
> 
> It reverts to using tf2onnx completly for the Keras model, which uses a different codepath and different optimizations which seem to make tensorrt happier.

Is the onnx model CHW format after using this method? so if I want to use it as a engine with tensorRT should I use is as a CHW or HWC?

**Comment by axeldavy** [comment]
I don't know. Isn't onnx HWC only ?

**Comment by Toku11** [comment]
No, in my view when you convert to onnx from HWC Tensorflow, It will transpose almost all layers to CHW except input, so if you use this model as a TensorRT model it will be a CHW model (internally for performance purposes) with HWC input. but I'm not really sure about it.

**Comment by kevinch-nv** [comment]
IIRC ONNX should specify CHW as the default, and TF2ONNX should be able to transpose HWC models to CHW when exporting to ONNX. 

@axeldavy are you still having this issue? Are you able to repro this on the latest TRT version?

**Comment by axeldavy** [comment]
@kevinch-nv Thanks for the follow-up. I do not use anymore the same codepath as I use tensorflow 2 now instead of Keras and tensorflow 1.
For this reason I cannot answer this question.

I would like to add thanks to the TensorRT team as TensorRT's onnx support is by far the best among what I could try. I'm sure the insane amount of ways to encode a given network with onnx makes it very hard to implement support. Even the reference onnx implementation crashes with my networks unless I give it very simple ones. Tensorflow doesn't even manage to import correctly its own onnx.

However there is still a lot left to be desired, and support for layers and onnx versions can be improved. This bug report is one example among many. Given slow/lack of evolution on my bug reports related to layer support, I stopped reported new ones.

I understand you are going to close this bug report, and that is the reasonable thing to do, given I don't have time to reproduce the error and no other user complained. In a better world, your team would have had time to look at it faster, and would make subreleases more often such that users can get fixes faster.


And since you are still reading, completly unrelated, but given tensor cores become so fast, bandwidth is more and more the limitation inside networks. It becomes even more important to have the activation function merged with the convolution operation. It would be a real great feature if TensorRT could merge other things than RELU, for example detecting patterns for all common activation functions (up to recent ones like swish or mish) and merge them. Because of that, I restrain myself to RELU, which is sad.

**Comment by kevinch-nv** [comment]
Thanks for your your thoughts. We've been trying to be better with monthly releases to provide support and bug fixes for some of the operators that we don't fully support. We are also continuously improving our graph optimizer so stay in tune for more optimizations in the future 😄 

I hope you continue to use TRT as your inference solution in the future, and feel free to open new issues you come across.
